dascore/proc/whiten.py

```diff
@@ -0,0 +1,210 @@
+       parameter.
+
+    3) Amplitude is NOT preserved
+
+
+    Example
+    -------
+
+    import matplotlib.pyplot as plt
+    import numpy as np
+    import numpy.fft as fft
+
+    import dascore as dc
+    from dascore.units import Hz
+
+
+    def plot_spectrum(x, T, ax, phase=False):
+    fftphase = np.angle(fft.fft(x))
+    fftsig = np.abs(fft.fft(x))
+    fftlen = fftsig.size
+    fftsig = fftsig[0 : int(fftlen / 2) + 1]
+    fftphase = fftphase[0 : int(fftlen / 2) + 1]
+    freqvec = np.linspace(0, 0.5 / T, fftsig.size)
+    if not phase:
+        ax.plot(freqvec, fftsig)
+        ax.set_xlabel("frequency [Hz]")
+        ax.set_ylabel("Amplitude (|H(w)|)")
+    else:
+        ax.plot(freqvec, fftphase)
+        ax.set_xlabel("frequency [Hz]")
+        ax.set_ylabel("Phase (radians)")
+
+
+    patch = dc.get_example_patch("dispersion_event")
+    patch = patch.resample(time=(200 * Hz))
+
+    filt_patch = patch.pass_filter(time=(5, 60))
+    white_patch = filt_patch.whiten(freq_range=[10, 50], freq_smooth_size=3)
+
+    fig, ((ax1, ax2), (ax3, ax4), (ax5, ax6)) = plt.subplots(3, 2, figsize=(10, 7))
+    ax1.plot(filt_patch.data[50, :])
+    ax1.set_title("Original data, distance = 50 m")
+    ax2.plot(white_patch.data[50, :])
+    ax2.set_title("Whitened data, distance = 50 m")
+
+    plot_spectrum(filt_patch.data[50, :], 1 / 200, ax3)
+    ax3.set_title("Original data, distance = 50 m")
+    plot_spectrum(white_patch.data[50, :], 1 / 200, ax4)
+    ax4.set_title("Whitened data, distance = 50 m")
+
+    plot_spectrum(filt_patch.data[50, :], 1 / 200, ax5, phase=True)
+    ax5.set_title("Original data, distance = 50 m")
+    plot_spectrum(white_patch.data[50, :], 1 / 200, ax6, phase=True)
+    ax6.set_title("Whitened data, distance = 50 m")
+    plt.tight_layout()
+    plt.show()
+
+    """
+    patch_cop = patch.convert_units(distance="m").transpose("distance", "time")
+    dt = 1.0 / get_dim_sampling_rate(patch_cop, "time")
+
+    if freq_range is None:
+        freq_range = [0, 1 / 2 / dt]
+    else:
+        if not np.size(freq_range) == 2:
+            msg = "Frequency range must include two values"
+            raise ParameterError(msg)
+
+        if freq_range[0] < 0 or freq_range[1] < 0:
+            msg = "Minimal and maximal frequencies have to be non-negative"
+            raise ParameterError(msg)
+
+        if freq_range[1] >= 0.5 / dt:
+            msg = "Frequency range exceeds Nyquist frequency"
+            raise ParameterError(msg)
+
+        if freq_range[0] >= freq_range[1]:
+            msg = "Frequency range must be increasing"
+            raise ParameterError(msg)
+
+    if freq_smooth_size is None:
+        freq_smooth_size = freq_range[1] - freq_range[0]
+    else:
+        if freq_smooth_size <= 0:
+            msg = "Frequency smoothing size must be positive"
+            raise ParameterError(msg)
+    if tukey_alpha is None:
+        tukey_alpha = 0.1
```

Comment:
Why not just set this default in the function signature? 

```python
def whiten(
    patch: PatchType,
    freq_range: tuple[float, float] | None = None,
    freq_smooth_size: float | None = None,
    tukey_alpha: float = 0.1,
) -> PatchType:
```

Reply:
Indeed.